### ISSUE TITLE
Bonus Beleg

### ISSUE THREAD

**Issue by KujaEx** [MattiWe/safe-secure-supergroup#11]
Wollt nur sagen, dass ich heut noch dran saß bei Task 1 alles hinzubekommen, dass es läuft. Hatte Eik angeschrieben, ob meine Lösung (trotz Funktionenübergabeparameteränderung) so okay ist. Als Antwort bekam ich:

"das kannst du machen, das klingt gut.
Mein Iterator ist ein Record der auch einen Access auf die von ihm iterierte Liste hatte, und kann First und Last von dieser auch ohne zusätzliche Übergabe ändern. "

Matti meinte ja, wenn man sich nicht 100%ig an die .ads hält, dann gibts Punktabzug. Weiß nicht, wie ich das aus Eiks Antwort interpretieren soll, ob ichs so lassen kann.
Hab jetzt aber glaub auch verstanden, wie Eik es implementiert hat (etwas umständlich, wenn man es mit Java oder C++ vergleichen würde). Ich könnte es genauso implementieren, aber das würde nochmal nen Tag wegnehmen, weil ich ca. 4 Funktionen umschreiben muss.

Soll wirklich bis Freitag der Beleg fertig sein? Das wird ne knappe Sache :/

**Comment by MattiWe** [comment]
Nach eingehender Betrachtung glaube ich auch nicht mehr daran, dass wir Freitag fertig werden könnten

**Comment by NaNaDi** [comment]
Heyho,
ich wollte mich dann mal um Task5 kümmern.
Wieso genau sollte das denn bis Freitag fertig sein?

**Comment by MattiWe** [comment]
nein, offiziell ist 10 März Abgabe. Ich wollte das nur mit new Woche Abstand vor der Prüfung fertig haben

**Comment by MattiWe** [comment]
Btw: Kevin kannst du die 2 noch machen? Ich hab da definitiv kb drauf, ich mach dann lieber die 3 :)

**Comment by KujaEx** [comment]
Sorry, ich hab von Parallelisierung und Hash Collision keinen Plan.
Ich kanns mir anschauen, aber wird nicht viel bei rauskommen.

Saß heute auch wieder an Task 1... und es ist einfach unmöglich, so wie es Eik als .ads vorgegeben hat. Man muss die List einfach in out setzen, da man sonst First und Last nicht ändern kann, besonders wenn sie zu Beginn Null sind. Falls Eik das wirklich so implementiert hat, dann hat das ziemlich wenig noch mit der Logik von einer normalen Double Linked List zu tun.
Ich schreib grad noch ein paar Tests um mehr Fälle abzudecken... das sind ja 10 Funktionen mit so vielen if-Bedingungen...

**Comment by MattiWe** [comment]
Zugegeben, bei der 4 ist die ADS auch ein Paradoxon ... Entweder ändert man die Spezifikation oder man ignoriert deutlich was die Spezifikation suggeriert was passieren soll

**Comment by NaNaDi** [comment]
Ich bin schon recht weit bei der Aufgabe 5. Wenn es euch doch nichts ausmacht, dass wir erst Anfang nächster Woche abgeben, könnte ich mich noch an der 2 versuchen. Parallelisierung habe ich die letzten Assignments jedes Mal gemacht und Hash Collision dürfte ich auch hinkriegen ;)

**Comment by KujaEx** [comment]
Saß heute etwas an Aufgabe 3... hier zwei Probleme:
1) Bei jeder Multiply-Operation (rekursiv) wird die Matrix in 4 gleich große teile geteilt. Dieses Matrix Slicing funktioniert nicht so richtig. Bezüglich Sub-Matritzen findet man zu Ada auch nicht viel im Internet...
2) Scheinbar kann man Strassen auch nicht 1:1 implementieren, weil Eik es wieder in einen Ring haben will mit nur + und * Operationen... fraglich ist, was dann aus der - Operation im Strassen-Algorithmus wird.

Wenn der rekursive Strassen-Algorithmus einmal steht, sollte die Parallelisierung kein Problem mehr sein, weil man dann einfach(?) vor jeden rekursiven Neu-Aufruf der Multiply Funktion einen neuen Thread erstellt.

btw: Eik ist grad in USA und hat Belege scheinbar noch nicht korrigiert... kommt nächste Woche Dienstag zurück.

**Comment by MattiWe** [comment]
Ich hab auch an der 3 gesessen und versucht das parsing hinzubekommen, aber das ist natürlich ada-style  mega kompliziert :(

**Comment by KujaEx** [comment]
@NaNaDi : Kannst ja die 2 anfangen, scheint aber auch nicht einfach zu sein...

@MattiWe : Hab jetzt bei 3 ne Slice und Merge Funktion gemacht für die C und M Matrizen und es werden keine Fehler mehr geworfen. Rein von der Logik her müssten sie funktionieren... nur die Main (wie du bereits in Mensa gesagt hast) müsste noch gemacht werden und wie die Subtraktion von Matritzen in dem Ring gehandhabt wird, da diese da nicht definiert ist (aber bei der "Sub" Funktionen werden zurzeit die Elemente einzeln miteinander substrahiert).

**Comment by MattiWe** [comment]
@KujaEx Wie brauchst du die Matrix indices? (1) für die Reihen und (2) für Zeilen oder andersrum?

**Comment by MattiWe** [comment]
@KujaEx Abgesehen davon wirft deine Multiply nen Stack Overflow .. so kann ich nicht arbeiten ;DD

**Comment by KujaEx** [comment]
Für den Algorithmus, müssen die Matrizen nen 2^n sein, also quadratisch mit ner Länge von 2,4,8 etc. ansonsten funktioniert das divide and conquer nicht :/
So wie ichs grad seh haben A und B ne Dimension von 6x6.

**Comment by MattiWe** [comment]
Dann sollten wir aber die exception werfen ... ich hab grade legit 2 Studen nach dem Fehler gesucht weil ich das in nem Task hatte. Hab dann das Armageddon eingebaut und habs dadurch nur gefunden ... 
Ich probiers nochmal mit 8 x 8 Matrizen

Nichtsdestotrotz fehlt noch das Tasking im Package ;)

**Comment by KujaEx** [comment]
Sorry... aber in der Funktion-Beschreibung steht nur, dass man ne Exception schmeißen soll, wenn A und B unterschiedlich groß sind.

Meinst du Parallelisieren? Davon hab ich keine Ahnung... man müsste halt jedes mal, wenn Multiply aufgerufen wird, dies auf einen neuen Thread machen (bzw. solang neue erzeugen, bis man n maximum erreicht hat, also mit hochzählen)

**Comment by MattiWe** [comment]
hab jetzt auf 8x8 matrizen umgestell und bekomme jetzt 
"raised CONSTRAINT_ERROR : strassen.adb:189 index check failed"

**Comment by KujaEx** [comment]
Wenn ich es starte kommt:
raised CONSTRAINT_ERROR : a-comlin.adb:65 explicit raise

Du hast die Exceptions umgestellt? Müsste da nicht trotzdem raise exception hin? So hat es auch Warnungen...
93 ist auch keine Primzahl...
Ansonsten wüsst ich auch nicht was gemeint ist...

**Comment by MattiWe** [comment]
Ja die warnungen sind ja egal. Und wenn ich 1031 nehme kommt das immernoch

**Comment by MattiWe** [comment]
Hab Strassen getestet -> der Algorithmus funktioniert, ABER er berechnert das Produkt der Transversen. 
Du musst das also schon noch in der Strassen.adb ändern mit den indices, ich tausche die Belegung der indices in der main wieder zurück zu (1) Reihen, (2) Spalten

**Comment by NaNaDi** [comment]
Heyho :)
Bei Task 5 kriege ich das Programm einfach nicht gescheit abgebrochen bei einem Timeout oder einer Nutzereingabe. Saß da jetzt schon relativ lange dran. Werde mich erstmal um Task 2 kümmern und dann nochmal darauf zurück kommen.

**Comment by MattiWe** [comment]
Kannst ja mal in meine 4 schaun, da hab ich das komplett generisch schon implementiert ;)

**Comment by NaNaDi** [comment]
Ooh wie cool! Das ist genau das was ich brauche! Baue es direkt mal ein, danke! :)

**Comment by NaNaDi** [comment]
@MattiWe Kann es sein, dass Du einen kleinen Fehler bei Deinem Time_Interrupted_Task hast? Bzw. was passiert wenn der Task fertig ist bevor die Zeit abgelaufen ist? Du wartest ja einfach bis Is_Interrupted_True auf True gesetzt wird und dann terminierst Du. Wenn der Task aber schon vorher fertig ist, kommt das ganze (zumindest bei mir) in eine Endlosschleife.

**Comment by MattiWe** [comment]
Bei mir terminiert alles regulär, habs grad nochmal ausprobiert.
